### PR TITLE
KAFKA-14451: Rack-aware consumer partition assignment for RangeAssignor (KIP-881)

### PR DESCRIPTION
**Pull request by rajinisivaram**
Best-effort rack alignment for range assignor when both consumer racks and partition racks are available with the protocol changes introduced in KIP-881. Rack-aware assignment is enabled by configuring `client.rack` for consumers. Balanced assignment per topic is prioritized over rack-alignment. For topics with equal partitions and the same set of subscribers, co-partitioning is prioritized over rack-alignment.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java]: > Topics with equal partition count and same set of subscribers prioritize co-partitioning guarantee over rack alignment.

I would like to ensure that we are on the same point on this point. My understanding is that the current implementation guarantees the co-partitioning iff the topics have the same number of partitions. Am I correct?

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java]: nit: Would it make sense to use `forEach` here? I always find the `getKey` and `getValue` annoying.

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java:151]: This makes sense. Unfortunately, we did not make the class final.

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java]: At first, it is not so clear why both `assignWithRackMatching` and `assignRanges` must be called. It may be worth adding a comment to explain the assignment logic here or some javadoc for `assignPartitions`.

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java]: It feels a bit weird to have this mapping here because it will be same in all `TopicAssignmentState` if the subscriptions are the same for all consumers. Is there a reason why you decided to have it here?

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java:193]: nit: I suppose that we could rename `i` to `p` and remove this one, isn't it?

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java:198]: What do we do with the partition if we don't find a consumer?

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java:205]: What do we do with the partition if we end up here?

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java:192]: I wonder if going over the partition is the best way. Imagine a case where the last partition ends up having only one rack available. This must be rare, I agree. However, you could end up in a situation where that partition could not be assigned to a consumer with the correct rack because all the consumers with that rack may have reached their maximum capacity.

When I was thinking about it, I was wondering if it would be better to start with the most constraint partitions: the partitions with only one rack available, then the partitions with two racks, etc. Did you consider something like this? There is perhaps a downside that I haven't though about. Once concern is that it may be less deterministic and the determinism is important.

File: clients/src/test/java/org/apache/kafka/clients/consumer/RangeAssignorTest.java
Comment (dajac) [inline clients/src/test/java/org/apache/kafka/clients/consumer/RangeAssignorTest.java]: nit: Should we inline `nonRackAwareAssignment` like for the other ones?

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (dajac) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: nit: I think that we usually put a space before `{`.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (dajac) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: When we have a consumer in the same zone as a leader, does the assignment algorithm guarantees that the leader in the same zone is selected all the time?

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (dajac) [inline core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: I was wondering if there is any value in having an integration test with FFF enabled as well. I am really not sure. What do you think?

**Review by dajac — COMMENTED** [review]
@rajinisivaram Thanks for the PR! I have started to dig into it and I left some comments/questions. It would be great if you could add somewhere a textual explanation of the full assignment algorithm because it was not easy to dive into the code without a high level understanding of the high level algorithm.

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java]: Yes, that is correct.

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java]: Done.

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java]: Added javadoc.

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java]: It was useful to have the consumer racks along with the partition racks for each topic for matching racks. But we don't need to populate it for each topic. I have removed this map and updated the list of consumers of the topic to include the rack.

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java:193]: Needed `p` since the lamdas required a final variable.

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java:198]: We fall-back to non-rack-aware assignment for all unassigned partitions, using the same logic as today.

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java:205]: As before, we fall-back to non-rack-aware assignment for all unassigned partitions after attempting rack-aware assignment.

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java:192]: I thought about these type of edge cases and felt that the logic is already much more complicated than it used to be. Handling edge cases would make it even more complex in terms of both implementation and testing. For example, in this particular scenario, we have n topics with p partitions and the n*p partitions could have replicas on different number of racks, but we want to co-partition. The current implementation works in a typical case where partition and consumer racks are uniform. We could extend further in follow-on PRs if we find that the edge cases are likely scenarios. What do you think?

File: clients/src/test/java/org/apache/kafka/clients/consumer/RangeAssignorTest.java
Comment (rajinisivaram) [reply clients/src/test/java/org/apache/kafka/clients/consumer/RangeAssignorTest.java]: Done.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (rajinisivaram) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: Done.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (rajinisivaram) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: No, we don't do any leader-specific assignment, we are assuming all replicas are equal.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (rajinisivaram) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: I wanted to put this test into a class that had FFF enabled, but I couldn't find any integration tests for FFF. So included here.

**Comment by rajinisivaram** [comment]
@dajac Thanks for the review. I have addressed the comments and left some questions.

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java:104]: @rajinisivaram I am trying to understand how we preserve the member order in the new algorithm. I am actually not sure that we preserve this property. It is important for when static member id are defined. Could we add a unit test which verifies that the assignment does not change when a member rejoins with a different member id but with the same static member id?

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java:192]: Yeah, I agree with you.

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java:104]: @dajac Thanks for the review. This `sort` is still done, I just moved it to TopicAssignmentState where the members are processed. All the tests in RangeAssignorTest are also run in rack-aware mode where all partitions have all racks to verify that the new algorithm gives the same result as the old algorithm. RangeAssignorTest.testStaticMemberRangeAssignmentPersistentAfterMemberIdChanges tests the scenario you described. I have added another version of the same test with a lower replication factor to verify the case where racks have a subset of partitions.

Comment (dajac) [reply clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java:104]: Thanks. I missed it in the `TopicAssignmentState`.

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java]: nit: I was wondering if using a `LinkedHashSet` would be better here because we have to remove partition from it in `onAssigned`. Did you consider this?

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java]: nit: I was wondering if we should directly limit based on `assignmentState.maxAssignable(consumer)` at this stage. For large groups with a large number of partitions, we don't really have to generate the full list as we limit it anyway right after.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (dajac) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: We have `FetchFromFollowerIntegrationTest`. Not sure if this what you were looking for.

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java]: nit: `final static` -> `static final` to be consistent with L98?

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java]: nit: Should we also use a set here as we remove the selected consumer from it?

**Review by dajac — COMMENTED** [review]
@rajinisivaram Thanks for the PR. I left a few nits. Otherwise, LGTM.

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java]: Makes sense, updated

File: clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/consumer/RangeAssignor.java]: Good idea, updated.

File: core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala
Comment (rajinisivaram) [reply core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala]: Ah, I missed that one, I think it is one of the new tests that was added when fixing FFF issues. I moved this test to that class. It uses 2 brokers instead of 3, but that seems ok for this test. Thanks.

**Comment by rajinisivaram** [comment]
@dajac Thanks for the reviews, I have addressed your comments.

**Review by dajac — APPROVED** [review]
LGTM, assuming tests look good. Thanks @rajinisivaram!

**Comment by rajinisivaram** [comment]
@dajac Thanks for the reviews. Test failures not related (have verified them locally), merging to trunk.

**Comment by twmb** [comment]
:wave: hi, seeing this after it's merged! Is the intent to create a new PR for making the sticky assignor rack aware? I see in the original PR #12914 that both range and sticky were improved. Thanks!

**Pull request merged**
